docs: added content for operations in AsyncAPI Document

BhaswatiRoy opened the pull request:
Fix: https://github.com/asyncapi/website/issues/1512

Content:

Meaning of operations
Steps for adding operations
Types of operations
Work: GSoD 2023 Issue

Continuation: https://github.com/asyncapi/website/pull/1777

BhaswatiRoy committed:
document adding-operations

TRohit20 commented on pages/docs/concepts/asyncapi-document/adding-operations.md:
```diff
@@ -0,0 +1,113 @@
+weight: 60
+---
+
+In a messaging system, the term "operations" refers to the various methods by which messages are exchanged between participants or components. 
```
In the introduction part of the document, it would be nice to include what operations also mean in AsyncAPI document, how they relate, how they work or used briefly. Doing so, the reader will have better understanding and more context

BhaswatiRoy replied:
added some more lined within the introduction

TRohit20 commented on pages/docs/concepts/asyncapi-document/adding-operations.md:
```diff
@@ -0,0 +1,113 @@
+
+In a messaging system, the term "operations" refers to the various methods by which messages are exchanged between participants or components. 
+
+## Operations Features
```
The document is about operations alone so it would give out the same meaning. Again, Keep the headings as concise yet informative as possible.
```suggestion
## Features
```

TRohit20 commented on pages/docs/concepts/asyncapi-document/adding-operations.md:
```diff
@@ -0,0 +1,113 @@
+
+## Operations Features
+
+- Operations describe the behaviors and capabilities of the messaging channels described in the document.
```
Be specific to avoid confusion. Document can be ambiguous or confusing for a newbie to comprehend or catch on to. 
```suggestion
- Operations describe the behaviors and capabilities of the messaging channels described in the AsyncAPI document.
```

TRohit20 commented on pages/docs/concepts/asyncapi-document/adding-operations.md:
```diff
@@ -0,0 +1,113 @@
+
+- The purpose of these operations is to provide a standardized means for describing the process of publishing, subscribing to, requesting, or replying to messages within the messaging system.
+
+## Steps to Add Operations
```
```suggestion
## Adding Operations
```

TRohit20 commented on pages/docs/concepts/asyncapi-document/adding-operations.md:
```diff
@@ -0,0 +1,113 @@
+
+## Steps to Add Operations
+
+For adding operations to an AsyncAPI document, we need to define them within the channels section of the document. The step-by-step guide on how to add operations to an AsyncAPI document is as follows - 
```
```suggestion
For adding operations to an AsyncAPI document, we need to define them within the channels section of the document. You can add operations to an AsyncAPI document as follows - 
```

TRohit20 commented on pages/docs/concepts/asyncapi-document/adding-operations.md:
```diff
@@ -0,0 +1,113 @@
+
+- Open your AsyncAPI document in a text editor or an AsyncAPI editor tool.
+
+- Locate the `channels` section in your AsyncAPI document. This section defines the messaging channels and their associated operations.
```
Again, 'this' is an antecedent remember?

```suggestion
- Locate the `channels` section in your AsyncAPI document. The `channels` section defines the messaging channels and their associated operations.
```

TRohit20 commented on pages/docs/concepts/asyncapi-document/adding-operations.md:
```diff
@@ -0,0 +1,113 @@
+
+- Within the `channels` section, add a new channel using the desired channel name. Channels are represented as properties within the `channels` object, and their names are usually formatted as paths or topics, depending on the messaging protocol you are using (e.g. `/users/{userId}/notifications` or `user.notifications`).
+
+- Inside the channel definition, specify the desired operations by creating sub-properties for each operation. Common operations include `publish`, `subscribe`, `publishSubscribe`, `requestReply`, and request.
```
```suggestion
- Inside the channel definition, specify the desired operations by creating sub-properties for each operation. Common operations include `publish`, `subscribe`, `publishSubscribe`, `requestReply`, and `request`.
```

TRohit20 commented on pages/docs/concepts/asyncapi-document/adding-operations.md:
```diff
@@ -0,0 +1,113 @@
+
+- Inside the channel definition, specify the desired operations by creating sub-properties for each operation. Common operations include `publish`, `subscribe`, `publishSubscribe`, `requestReply`, and request.
+
+- For each operation, define its details using the appropriate keywords and properties. These details include the operation type, payload schema, headers, bindings, and other relevant information.
```
```suggestion
- For each operation, define its details using the appropriate keywords and properties. These details typically include the operation type, payload schema, headers, bindings, and other relevant information.
```

TRohit20 commented on pages/docs/concepts/asyncapi-document/adding-operations.md:
````diff
@@ -0,0 +1,113 @@
+
+- For each operation, define its details using the appropriate keywords and properties. These details include the operation type, payload schema, headers, bindings, and other relevant information.
+
+```mermaid
````
Would be nice to have a sentence introducing the diagram and what it illustrates?

BhaswatiRoy replied:
added a heading to the mermaid diagram describing what it is actually showing

TRohit20 commented on pages/docs/concepts/asyncapi-document/adding-operations.md:
````diff
@@ -0,0 +1,113 @@
+    class A,B,C,D,E,F,G,H,I,J labelStyle;
+```
+
+## Operations Types
````
```suggestion
## Types
```

TRohit20 commented on pages/docs/concepts/asyncapi-document/adding-operations.md:
```diff
@@ -0,0 +1,113 @@
+
+## Operations Types
+
+The operations of the AsyncAPI document are divided into the following different categories based on the purpose they serve -
```
```suggestion
The operations of the AsyncAPI document are divided into the following categories based on the purpose they serve -
```

TRohit20 commented on pages/docs/concepts/asyncapi-document/adding-operations.md:
````diff
@@ -0,0 +1,113 @@
+    class A,B,C,D,E,F,G,H,I,J labelStyle;
+```  
+
+- <b> Request-Reply: </b> RequestReply establishes a pattern of request-response communication between components. The client can send a request message, and the server will respond with a corresponding reply message. Typically, this operation is used for synchronous communication in which the client anticipates a response from the server.
````
```suggestion
- <b> Request-Reply: </b> Request/Reply establishes a pattern of request-response communication between components. The client can send a request message, and the server will respond with a corresponding reply message. Typically, this operation is used for synchronous communication in which the client anticipates a response from the server.
```

TRohit20 reviewed (changes requested):
Use backticks wherever required throughout the document.

BhaswatiRoy committed:
Update pages/docs/concepts/asyncapi-document/adding-operations.md

Co-authored-by: Rohit <108233235+TRohit20@users.noreply.github.com>

BhaswatiRoy committed:
Update pages/docs/concepts/asyncapi-document/adding-operations.md

Co-authored-by: Rohit <108233235+TRohit20@users.noreply.github.com>

BhaswatiRoy committed:
Update pages/docs/concepts/asyncapi-document/adding-operations.md

Co-authored-by: Rohit <108233235+TRohit20@users.noreply.github.com>

BhaswatiRoy committed:
Update pages/docs/concepts/asyncapi-document/adding-operations.md

Co-authored-by: Rohit <108233235+TRohit20@users.noreply.github.com>

BhaswatiRoy committed:
Update pages/docs/concepts/asyncapi-document/adding-operations.md

Co-authored-by: Rohit <108233235+TRohit20@users.noreply.github.com>

BhaswatiRoy committed:
Update pages/docs/concepts/asyncapi-document/adding-operations.md

Co-authored-by: Rohit <108233235+TRohit20@users.noreply.github.com>

BhaswatiRoy committed:
Update pages/docs/concepts/asyncapi-document/adding-operations.md

Co-authored-by: Rohit <108233235+TRohit20@users.noreply.github.com>

BhaswatiRoy committed:
Update pages/docs/concepts/asyncapi-document/adding-operations.md

Co-authored-by: Rohit <108233235+TRohit20@users.noreply.github.com>

BhaswatiRoy committed:
Update pages/docs/concepts/asyncapi-document/adding-operations.md

Co-authored-by: Rohit <108233235+TRohit20@users.noreply.github.com>

BhaswatiRoy committed:
Update pages/docs/concepts/asyncapi-document/adding-operations.md

Co-authored-by: Rohit <108233235+TRohit20@users.noreply.github.com>

derberg reviewed (commented):
I like the intro and features a lot, well explained, but for the rest I stopped review as you need to go through it again and make sure it is really about v3, because `operations` are no longer under `channels` in v3, these are separate objects in AsyncAPI document, on the same level, on root. And you link them by referencing inside operation what channel it relates to

https://www.asyncapi.com/docs/reference/specification/v3.0.0-next-major-spec.14#operationsObject

```
onUserSignUp:
  title: User sign up
  summary: Action to sign a user up.
  description: A longer description
  channel:
    $ref: '#/channels/userSignup'
```

BhaswatiRoy committed:
updated the adding operations sub section

BhaswatiRoy committed:
updated adding operations subsection

BhaswatiRoy committed:
removed unwanted characters

BhaswatiRoy commented:
Hello @derberg I have updated the contents as per your technical review, please let me know if you want me to update any thing more.

derberg commented on pages/docs/concepts/asyncapi-document/adding-operations.md:
```diff
@@ -0,0 +1,132 @@
+
+- In a messaging channel, an operation represents a particular action or interaction that can be performed. 
+
+- The purpose of these operations is to provide a standardized means for describing the process of publishing, subscribing to, requesting, or replying to messages within the messaging system.
```
```suggestion
- The purpose of these operations is to provide a standardized means for describing the process of sending, receiving from, requesting, or replying to messages within the messaging system.
```

derberg commented on pages/docs/concepts/asyncapi-document/adding-operations.md:
```diff
@@ -0,0 +1,132 @@
+
+## Adding Operations
+
+`operations` are no longer under `channels` in AsyncAPI Spec 3.0.0, instead, `operations` are separate objects in the AsyncAPI document on the same level. 
```
```suggestion
`operations` are separate objects in the AsyncAPI document on the same level. 
```

please write in the way what is possible now, and not how it worked before

derberg commented on pages/docs/concepts/asyncapi-document/adding-operations.md:
```diff
@@ -0,0 +1,132 @@
+## Adding Operations
+
+`operations` are no longer under `channels` in AsyncAPI Spec 3.0.0, instead, `operations` are separate objects in the AsyncAPI document on the same level. 
+`channels` can be linked within `operations` by referencing them within the `channels`, just like the following example -
```
where is this `-` suppose to point? did you forget to add example or?

BhaswatiRoy replied:
corrected the sequence of the example below the statement

derberg commented on pages/docs/concepts/asyncapi-document/adding-operations.md:
```diff
@@ -0,0 +1,132 @@
+
+`operations` are no longer under `channels` in AsyncAPI Spec 3.0.0, instead, `operations` are separate objects in the AsyncAPI document on the same level. 
+`channels` can be linked within `operations` by referencing them within the `channels`, just like the following example -
+For adding operations to an AsyncAPI document, we need to define them within the channels section of the document. You can add operations to an AsyncAPI document as follows - 
```
> For adding operations to an AsyncAPI document, we need to define them within the channels section of the document

no, operations are not part of channels

BhaswatiRoy replied:
removed that line as it was unneccesary, it possibly was a residue from the old draft

derberg commented on pages/docs/concepts/asyncapi-document/adding-operations.md:
````diff
@@ -0,0 +1,132 @@
+For adding operations to an AsyncAPI document, we need to define them within the channels section of the document. You can add operations to an AsyncAPI document as follows - 
+
+```
+onUserSignUp:
````
missing `action`

BhaswatiRoy replied:
added `action` in the example

derberg commented on pages/docs/concepts/asyncapi-document/adding-operations.md:
````diff
@@ -0,0 +1,132 @@
+      ack: false
+  traits:
+    - $ref: '#/components/operationTraits/kafka'
+```
````
please do not duplicate what is already in the spec document, better just link to reference docs

maybe you wanna follow the approach other writers did, diagram representation of where operation is and what are the required fields (at the beginning of the document)

BhaswatiRoy replied:
referenced the parts of the actual spec 3 document - https://v3.asyncapi.com/docs/reference/specification/v3.0.0-next-major-spec.12#operationsObject and added a mermaid diagram to brief the field names is added in the beginning of the document.

derberg commented on pages/docs/concepts/asyncapi-document/adding-operations.md:
````diff
@@ -0,0 +1,132 @@
+    - $ref: '#/components/operationTraits/kafka'
+```
+
+## Types
````
what is the purpose of this section and where did you get this content from? in v3 there is no longer publishing and subscribing - and if it comes to request/reply, there will be separate concept document that you should just link to

maybe instead of Types you mean actions? `send` and `receive` where `send` means that application describe in AsyncAPI document "sends" a message and `receive` means it is "receiving" a message

BhaswatiRoy replied:
since this section is no more relevant to v3 I have removed that section. `send` and `receive` is already defined in the table of spec3 doc - https://v3.asyncapi.com/docs/reference/specification/v3.0.0-next-major-spec.12#operationsObject
which I already linked in the beginning of the document

BhaswatiRoy committed:
Update pages/docs/concepts/asyncapi-document/adding-operations.md

Co-authored-by: Lukasz Gornicki <lpgornicki@gmail.com>

BhaswatiRoy committed:
Update pages/docs/concepts/asyncapi-document/adding-operations.md

Co-authored-by: Lukasz Gornicki <lpgornicki@gmail.com>

BhaswatiRoy committed:
updated the changes including diagrams, and addition of content

BhaswatiRoy committed:
Merge branch 'gsod-work-1' of https://github.com/BhaswatiRoy/website into gsod-work-1

BhaswatiRoy committed:
updated changes

BhaswatiRoy commented:
Thanks @derberg for the re-review, I have updated all the changes as per your comments. Do let me know if I still missed anything and need to add/remove anything.

derberg commented on pages/docs/concepts/asyncapi-document/adding-operations.md:
```diff
@@ -0,0 +1,63 @@
+
+## Adding Operations
+
+`operations` are no longer under `channels` in AsyncAPI Spec 3.0.0, instead, `operations` are separate objects in the AsyncAPI document on the same level. 
```
```suggestion
`operations` are separate objects in the AsyncAPI document on the root level together with `channels` and other objects. 
```

let's not mention what was it like prior 3.0

derberg replied:
do we even need this sentence? you already wrote `Operations can be defined as an independent object in the AsyncAPI document` in previous section

derberg commented on pages/docs/concepts/asyncapi-document/adding-operations.md:
```diff
@@ -0,0 +1,63 @@
+## Adding Operations
+
+`operations` are no longer under `channels` in AsyncAPI Spec 3.0.0, instead, `operations` are separate objects in the AsyncAPI document on the same level. 
+`channels` can be linked within `operations` by referencing them within the `channels`, just like the following example - 
```
it's not that it can, it must be, `channel` reference is required in operation

```suggestion
Operation must specify on what channel it is performed. You do it by referencing the `channel` with `$ref`, just like in the following example:
```

derberg commented on pages/docs/concepts/asyncapi-document/adding-operations.md:
```diff
@@ -0,0 +1,63 @@
+Operations can be defined as an independent object in the AsyncAPI document. Operations have the following components for it's definition. More information about each field names that are used to define operations can be found [here](https://v3.asyncapi.com/docs/reference/specification/v3.0.0-next-major-spec.12#operationObject). 
+Additionally, an example to show the usage of each field names in defining operations can be found [here](https://v3.asyncapi.com/docs/reference/specification/v3.0.0-next-major-spec.12#operationsObject).
+`operations` are separate objects in the AsyncAPI document on the same level. 
+`channels` can be linked within `operations` by referencing them within the `channels`, just like the following example -
```
you have the same in next section

BhaswatiRoy replied:
sorry, this line was supposed to be removed in the current draft, but for some glitch, it was not removed when I pushed the changes to github this time

BhaswatiRoy replied:
I have removed the line and pushed the changes again

derberg commented on pages/docs/concepts/asyncapi-document/adding-operations.md:
```diff
@@ -0,0 +1,63 @@
+style G fill:#01A9DB,stroke:#333,stroke-width:2px
+style H fill:#F781F3,stroke:#333,stroke-width:2px
+
+  A[Fields required to define AsyncAPI Operations] -->|1| B[Summary]
```
not true, only `action` and `channel` are required

BhaswatiRoy replied:
> not true, only `action` and `channel` are required

![image](https://github.com/asyncapi/website/assets/78029145/906c22cb-c58c-4012-ba30-4ee3486f0f62)

Should I divide the diagram for both required and optional fields for operations?

BhaswatiRoy replied:
divided the diagram and aligned it with other tech writers' mermaid diagrams

![image](https://github.com/asyncapi/website/assets/78029145/68a25c95-db13-47e7-aca0-e517c4070b30)

BhaswatiRoy replied:
I can see Rohit has used only 2 colors for the diagram nodes, waiting for Mahfuza to let me know if the color is used is bipartite or we can use more colors for the same

BhaswatiRoy replied:
got the diagram reviewed by @alequetzalli, and implemented her suggestions as per the style guide

derberg commented on pages/docs/concepts/asyncapi-document/adding-operations.md:
````diff
@@ -0,0 +1,63 @@
+The following diagram briefs the important field names that are frequently used to define AsyncAPI operations in Spec 3.0.0 -
+
+```mermaid
+flowchart TD
````
please align the style of the diagram with rest of writers, like Rohit or Mahfuza

BhaswatiRoy replied:
sure, I have dropped both Rohit & Mahfuza a text, for a sample of their writing style so that I can align my work with their.

BhaswatiRoy replied:
updated mermaid diagram code as per Rohit's diagram code

derberg commented on pages/docs/concepts/asyncapi-document/adding-operations.md:
```diff
@@ -0,0 +1,63 @@
+
+## Defining Operations
+
+Operations can be defined as an independent object in the AsyncAPI document. Operations have the following components for it's definition. More information about each field names that are used to define operations can be found [here](https://v3.asyncapi.com/docs/reference/specification/v3.0.0-next-major-spec.12#operationObject). 
```
> Operations have the following components for it's definition

sorry if I judge too quickly and draw false conclusions, but looking at this round of review, you did not really read what you wrote before submitting for another review round. Please make sure that what you write is accurate and ready for review (relating also to other comments where I noticed you just copy over stuff without making sure what you do)

BhaswatiRoy replied:
Sorry for the previous miscommunication. I have been trying to use GitHub Desktop lately after suggestions from other contributors, and I am not very good at using it, which is why I am seeing that some of my commits are not pushed into the repository properly especially the deletion of lines. I deleted all those lines that you pointed out in my VS Code editor, but for some reason, the changes are not showing up here.

I am extremely sorry for the inconvenience, and I will be more careful in the future.
I have made the required changes and also updated the contents with some details about the field names in adding operations. I have ensured all changes are pushed correctly this time.

BhaswatiRoy committed:
Update pages/docs/concepts/asyncapi-document/adding-operations.md

Co-authored-by: Lukasz Gornicki <lpgornicki@gmail.com>

BhaswatiRoy committed:
Update pages/docs/concepts/asyncapi-document/adding-operations.md

Co-authored-by: Lukasz Gornicki <lpgornicki@gmail.com>

BhaswatiRoy committed:
removed duplicate lines

this line was supposed to be removed from the old draft but for some reason, it still wasn't removed when I pushed the changes to github.

BhaswatiRoy committed:
aligned mermaid diagram code with other writers

BhaswatiRoy committed:
updated mermaid diagram to align with the style guide

BhaswatiRoy committed:
added field name details in operations

TRohit20 commented on pages/docs/concepts/asyncapi-document/adding-operations.md:
```diff
@@ -0,0 +1,133 @@
+
+- In a messaging channel, an operation represents a particular action or interaction that can be performed. 
+
+- The purpose of these operations is to provide a standardized means for describing the process of sending, receiving from, requesting, or replying to messages within the messaging system.
```
```suggestion
In a AsyncAPI document, Operations describe the behaviors and capabilities of the messaging channels described in the AsyncAPI document. In a messaging channel, an `operation` represents a particular action or interaction that can be performed. The purpose of operations is to provide a standardized means for describing the process of sending, receiving from, requesting, or replying to messages within the messaging system. 
```

TRohit20 commented on pages/docs/concepts/asyncapi-document/adding-operations.md:
```diff
@@ -0,0 +1,133 @@
+Operations can be defined as an independent object in the AsyncAPI document. Operations have the following components for their definition. More information about each field name that is used to define operations can be found [here](https://v3.asyncapi.com/docs/reference/specification/v3.0.0-next-major-spec.12#operationObject). 
+Additionally, an example to show the usage of each field name in defining operations can be found [here](https://v3.asyncapi.com/docs/reference/specification/v3.0.0-next-major-spec.12#operationsObject).
+
+The following diagram briefs the important field names that are frequently used to define AsyncAPI operations in Spec 3.0.0 -
```
```suggestion
The following diagram briefs the some field names that are frequently used to define operations in a AsyncAPI document:
```

TRohit20 reviewed (changes requested):
Left couple of suggestions, please update as you see fit.

BhaswatiRoy committed:
Update pages/docs/concepts/asyncapi-document/adding-operations.md

Co-authored-by: Rohit <108233235+TRohit20@users.noreply.github.com>

BhaswatiRoy committed:
Update pages/docs/concepts/asyncapi-document/adding-operations.md

Co-authored-by: Rohit <108233235+TRohit20@users.noreply.github.com>

BhaswatiRoy commented:
> Left couple of suggestions, please update as you see fit.

Thanks for the review, updated the suggestions!
